dll/win32/kernel32/client/dep.c

```diff
@@ -0,0 +1,147 @@
+/*
+ * PROJECT:         ReactOS system libraries
+ * LICENSE:         GPL-2.0-or-later (https://spdx.org/licenses/GPL-2.0-or-later)
+ * FILE:            dll/win32/kernel32/client/dep.c
+ * PURPOSE:         ReactOS Data Execution Prevention (DEP) functions
+ * COPYRIGHT:       Copyright 2021 Oleg Dubinskiy (oleg.dubinskij2013@yandex.ua)
+ */
+
+/* INCLUDES *******************************************************************/
+
+#include <k32.h>
+
+#define NDEBUG
+#include <debug.h>
+
+/* PUBLIC FUNCTIONS ***********************************************************/
+
+/**
+ * @brief Retrieves the data execution prevention (DEP) and thunk emulation (DEP-ATL)
+ * settings for the process.
+ * 
+ * On Windows XP SP3, the current process is handled.
+ * On Vista and newer, the process specified in hProcess parameter is handled.
+ * 
+ * @param [in] hProcess
+ * A handle to a process with PROCESS_QUERY_INFORMATION privilege.
+ * Ignored on Windows XP SP3.
+ * 
+ * @param [out] lpFlags
+ * Pointer to a variable that receives the flags:
+ * 0 - DEP is disabled.
+ * PROCESS_DEP_ENABLE - DEP is enabled.
+ * PROCESS_DEP_DISABLE_ATL_THUNK_EMULATION - DEP-ATL is disabled.
+ * 
+ * @param [out] lpPermanent
+ * Pointer to a variable that specifies whether DEP is enabled or disabled permanently.
+ * 
+ * @return
+ * TRUE on success, FALSE on failure.
+ * 
+ * @remarks
+ * Supported for 32-bit processes only. Returns ERROR_NOT_SUPPORTED for 64-bit processes.
```

Comment:
The 64bit behavior looks missing for the 3 functions, isn't it?

Reply:
Yes, seems to be. But regarding `GetSystemDEPPolicy`, MSDN says the following:
> If the system DEP policy is OptIn or OptOut, DEP can be selectively enabled or disabled for the current process by calling the SetProcessDEPPolicy function. This function works only for 32-bit processes.

It's not completely understandable, which function is meant in this context: `GetProcessDEPPolicy` or `GetSystemDEPPolicy`? Or both? :thinking:

Reply:
In this context the MSDN documentation states that setting DEP onto the current calling process with `SetProcessDEPPolicy` only works if the current calling process itself is a 32-bit process. Though indeed the documentation doesn't explicitly say how the function actually doesn't work so I'd suggest that you write an APITEST to testify the behavior.

Reply:
Btw, we already seem to have an appropriate winetest for that: https://git.reactos.org/?p=reactos.git;a=blob;f=modules/rostests/winetests/kernel32/virtual.c;hb=f028ca5b083524cc85fd75467b6f8db4e37f0164.

Reply:
That winetest barely tests anything about DEP implementation of Windows. So do what I've told you to do, implement a testcase on your own.